File: app/src/main/java/com/decentstorage/app/erasure/GF256.kt
package com.decentstorage.app.erasure

/**
 * Aritmética em GF(256) — corpo finito usado pelo Reed-Solomon.
 * Polinômio primitivo: x^8 + x^4 + x^3 + x^2 + 1 (0x11d), gerador = 2.
 * Port direto de gf256.js (mesma tabela, mesmo resultado byte a byte).
 */
object GF256 {
    private val EXP = IntArray(512)
    private val LOG = IntArray(256)

    init {
        var x = 1
        for (i in 0 until 255) {
            EXP[i] = x
            LOG[x] = i
            x = x shl 1
            if (x and 0x100 != 0) x = x xor 0x11d
        }
        for (i in 255 until 512) EXP[i] = EXP[i - 255]
    }

    fun add(a: Int, b: Int): Int = a xor b

    fun mul(a: Int, b: Int): Int {
        if (a == 0 || b == 0) return 0
        return EXP[LOG[a] + LOG[b]]
    }

    fun div(a: Int, b: Int): Int {
        if (a == 0) return 0
        require(b != 0) { "Divisão por zero em GF(256)" }
        return EXP[(LOG[a] - LOG[b] + 255) % 255]
    }

    fun pow(a: Int, n: Int): Int {
        if (n == 0) return 1
        if (a == 0) return 0
        return EXP[(LOG[a] * n) % 255]
    }

    fun inv(a: Int): Int {
        require(a != 0) { "Zero não tem inverso em GF(256)" }
        return EXP[255 - LOG[a]]
    }
}

File: sever/gateway/gf256.js
// Porta 1:1 de app/src/main/java/.../erasure/GF256.kt
// Tem que bater byte a byte com o Kotlin, senão o decode RS não fecha com o que o app gerou.

const EXP = new Int32Array(512);
const LOG = new Int32Array(256);

(function init() {
  let x = 1;
  for (let i = 0; i < 255; i++) {
    EXP[i] = x;
    LOG[x] = i;
    x = x << 1;
    if (x & 0x100) x ^= 0x11d;
  }
  for (let i = 255; i < 512; i++) EXP[i] = EXP[i - 255];
})();

function mul(a, b) {
  if (a === 0 || b === 0) return 0;
  return EXP[LOG[a] + LOG[b]];
}

function inv(a) {
  if (a === 0) throw new Error('Zero não tem inverso em GF(256)');
  return EXP[255 - LOG[a]];
}

function pow(a, n) {
  if (n === 0) return 1;
  if (a === 0) return 0;
  return EXP[(LOG[a] * n) % 255];
}

module.exports = { mul, inv, pow };
